paddle/fluid/imperative/reducer.cc

```diff
@@ -0,0 +1,300 @@
+// Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/imperative/reducer.h"
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include "paddle/fluid/framework/data_type.h"
+
+namespace paddle {
+namespace imperative {
+
+std::shared_ptr<Reducer> Reducer::s_instance_ = NULL;
+
+Reducer::Reducer(const std::vector<std::shared_ptr<imperative::VarBase>> &vars,
+                 const std::vector<std::vector<size_t>> &group_indices,
+                 std::shared_ptr<imperative::ParallelContext> parallel_ctx)
+    : vars_(vars), group_indices_(group_indices), parallel_ctx_(parallel_ctx) {
+  // parallel_ctx->Print_ParallelStrategy();
+  VLOG(3) << "Start construct the Reducer ...";
+  // initialize groups
+  initialize_groups(group_indices);
+
+  // initialize varname2index_
+  {
+    for (size_t group_index = 0; group_index < group_indices.size();
+         ++group_index) {
+      for (size_t var_index = 0; var_index < group_indices[group_index].size();
+           ++var_index) {
+        size_t index = group_indices[group_index][var_index];
+        const std::string &var_name = vars_[index]->GradVarName();
+        varname2index_[var_name] = VariableIndex{
+            .group_index = group_index, .variable_index = var_index,
+        };
+      }
+    }
+  }
+}
+
+void Reducer::initialize_groups(
```

Comment:
function name should comply with the conventions: (InitializeGroups here), the same as following several functions

Reply:
ok,fix it